x-pack/plugins/observability_ai_assistant/server/service/kb_service/index.ts

```diff
@@ -0,0 +1,294 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { errors } from '@elastic/elasticsearch';
+import type { QueryDslTextExpansionQuery } from '@elastic/elasticsearch/lib/api/types';
+import { serverUnavailable } from '@hapi/boom';
+import type { ElasticsearchClient } from '@kbn/core/server';
+import type { Logger } from '@kbn/logging';
+import { TaskManagerStartContract } from '@kbn/task-manager-plugin/server';
+import pLimit from 'p-limit';
+import pRetry from 'p-retry';
+import { INDEX_QUEUED_DOCUMENTS_TASK_ID, INDEX_QUEUED_DOCUMENTS_TASK_TYPE } from '..';
+import type { KnowledgeBaseEntry } from '../../../common/types';
+import type { ObservabilityAIAssistantResourceNames } from '../types';
+import { getAccessQuery } from '../util/get_access_query';
+
+interface Dependencies {
+  esClient: ElasticsearchClient;
+  resources: ObservabilityAIAssistantResourceNames;
+  logger: Logger;
+  taskManagerStart: TaskManagerStartContract;
+}
+
+const ELSER_MODEL_ID = '.elser_model_1';
+
+function throwKnowledgeBaseNotReady(body: any) {
+  throw serverUnavailable(`Knowledge base is not ready yet`, body);
+}
+
+export class KnowledgeBaseService {
+  private hasSetup: boolean = false;
+
+  private entryQueue: KnowledgeBaseEntry[] = [];
+
+  constructor(private readonly dependencies: Dependencies) {
+    this.ensureTaskScheduled();
+  }
+
+  private ensureTaskScheduled() {
+    this.dependencies.taskManagerStart
+      .ensureScheduled({
+        taskType: INDEX_QUEUED_DOCUMENTS_TASK_TYPE,
+        id: INDEX_QUEUED_DOCUMENTS_TASK_ID,
+        state: {},
+        params: {},
+        schedule: {
+          interval: '1h',
+        },
+      })
+      .then(() => {
+        this.dependencies.logger.debug('Scheduled document queue task');
+        return this.dependencies.taskManagerStart.runSoon(INDEX_QUEUED_DOCUMENTS_TASK_ID);
+      })
+      .then(() => {
+        this.dependencies.logger.debug('Document queue task ran');
+      })
+      .catch((err) => {
+        this.dependencies.logger.error(`Failed to schedule document queue task`);
+        this.dependencies.logger.error(err);
+      });
+  }
+
+  async processQueue() {
+    if (!this.entryQueue.length) {
+      return;
+    }
+
+    if (!(await this.status()).ready) {
+      this.dependencies.logger.debug(`Bailing on document queue task: KB is not ready yet`);
+      return;
+    }
+
+    this.dependencies.logger.debug(`Processing document queue`);
+
+    this.hasSetup = true;
+
+    this.dependencies.logger.info(`Indexing ${this.entryQueue.length} queued entries into KB`);
+    const limiter = pLimit(5);
+
+    const entries = this.entryQueue.concat();
+
+    await Promise.all(
```

Comment:
Not clear to me how many entries there could be, good to see `pLimit()` used :-)

With `pLimit()`, we run the risk of the `Promise.all()` taking longer, running the risk of the task timing out.  So we may want a `cancel()` method on the task; see https://github.com/elastic/kibana/blob/main/x-pack/plugins/task_manager/README.md#task-definitions .  That method would do something to force the `summarize()` method to fail / return early.

Could be done as a follow-up.